{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "from core.utils import Tibanna\n",
    "from dcicutils import ff_utils\n",
    "from core.utils import run_workflow\n",
    "from datetime import datetime\n",
    "from core.wfr import *\n",
    "\n",
    "env = 'data'\n",
    "tibanna = Tibanna(env=env)\n",
    "ff = ff_utils.fdn_connection(key={\"default\" : tibanna.ff_keys})\n",
    "exclude_miseq = True"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "41 total number of sets\n",
      "0 sets completed\n",
      "\n",
      "1 4DNESUZ3Y5GY\n",
      "None human\n",
      "4DNEXE23WWO1 part1 complete\n",
      "4DNEXG6HJJ27 part1 complete\n",
      "\n",
      "2 4DNESIV168N6\n",
      "None human\n",
      "4DNEXVRTS3VG part1 complete\n",
      "4DNEXELBLZ37 part1 complete\n",
      "\n",
      "3 4DNESXZ2IL4M\n",
      "None human\n",
      "4DNEXKP295JF part1 complete\n",
      "\n",
      "4 4DNESJOH66ZZ\n",
      "None human\n",
      "4DNEXYX58PCM part1 complete\n",
      "\n",
      "5 4DNESQ3ZDLZM\n",
      "None human\n",
      "4DNEXCTB91SG part1 complete\n",
      "\n",
      "6 4DNES98TS7X2\n",
      "None human\n",
      "4DNEXXIXTTXI part1 complete\n",
      "\n",
      "7 4DNESX6MGDN2\n",
      "None human\n",
      "4DNEXLVWLA33 part1 complete\n",
      "\n",
      "8 4DNES6O1SJ4O\n",
      "None human\n",
      "4DNEX1BWBWFA part1 complete\n",
      "\n",
      "9 4DNESWS6UT74\n",
      "None human\n",
      "4DNEXZPBA82F part1 complete\n",
      "\n",
      "10 4DNESBAF7EW5\n",
      "None human\n",
      "4DNEXDWEK1ME part1 complete\n",
      "\n",
      "11 4DNES9OZNM6T\n",
      "None human\n",
      "4DNEXTM3PBS3 part1 complete\n",
      "\n",
      "12 4DNESDFQDLQJ\n",
      "None human\n",
      "4DNEX8K5TT7Y part1 complete\n",
      "\n",
      "13 4DNES3DSRT51\n",
      "None human\n",
      "4DNEXS77PTUK part1 complete\n",
      "\n",
      "14 4DNES7OJ7U7X\n",
      "None human\n",
      "4DNEXQ4EXW3K part1 complete\n",
      "\n",
      "15 4DNESLS88OS7\n",
      "None human\n",
      "4DNEXEQB7RNA part1 complete\n",
      "\n",
      "16 4DNESDUJHQ6J\n",
      "None human\n",
      "4DNEX5M9XSGK part1 complete\n",
      "\n",
      "17 4DNESCBR4664\n",
      "None human\n",
      "4DNEXKCP8338 part1 complete\n",
      "\n",
      "18 4DNESJNUUJ66\n",
      "None human\n",
      "4DNEXL6WTINA part1 complete\n",
      "\n",
      "19 4DNES5PW6NSG\n",
      "None human\n",
      "4DNEXDRZLEQV part1 complete\n",
      "\n",
      "20 4DNESSV33VOL\n",
      "None human\n",
      "4DNEX63QU5GZ part1 complete\n",
      "4DNEXTG6XPXD part1 complete\n",
      "\n",
      "21 4DNESH4XLJCW\n",
      "None human\n",
      "4DNEXR87VFPR part1 complete\n",
      "4DNEXRK56264 part1 complete\n",
      "\n",
      "22 4DNES99LXRYK\n",
      "None human\n",
      "4DNEXWNB33S2 part1 complete\n",
      "4DNEX25G1FKX part1 complete\n",
      "\n",
      "23 4DNES1P18J2X\n",
      "None human\n",
      "4DNEXAM4T83J part1 complete\n",
      "4DNEXBL1Y5N4 part1 complete\n",
      "\n",
      "24 4DNESLC8TDK4\n",
      "None human\n",
      "4DNEX6SUB7CY part1 complete\n",
      "4DNEXAPQBU2M part1 complete\n",
      "\n",
      "25 4DNESSYKYIK3\n",
      "None human\n",
      "4DNEXLRK3X6D part1 complete\n",
      "4DNEXVQ9XCJ5 part1 complete\n",
      "\n",
      "26 4DNES1GSPUT8\n",
      "None human\n",
      "4DNEXPCVJ25I part1 complete\n",
      "4DNEXKWN7CZT part1 complete\n",
      "\n",
      "27 4DNES9YA22WT\n",
      "None human\n",
      "4DNEXBCDJFX9 part1 complete\n",
      "4DNEXXZUM1J1 part1 complete\n",
      "\n",
      "28 4DNES1PHNYOP\n",
      "None mouse\n",
      "4DNEX14L7KXP part1 complete\n",
      "4DNEXTTY1KTG part1 complete\n",
      "\n",
      "29 4DNESWPIPVZT\n",
      "None mouse\n",
      "4DNEX91MUU81 part1 complete\n",
      "4DNEXWTU45NI part1 complete\n",
      "\n",
      "30 4DNEST651P8O\n",
      "None human\n",
      "4DNEXK7XIMBW part1 complete\n",
      "4DNEXB7RB9KT part1 complete\n",
      "\n",
      "31 4DNESJQH3HJ9\n",
      "None human\n",
      "4DNEXUFNTG8J part1 complete\n",
      "4DNEX9GKA2OI part1 complete\n",
      "\n",
      "32 4DNES674QWXK\n",
      "None human\n",
      "4DNEX9CR3FWQ part1 complete\n",
      "4DNEXCDJJVUR part1 complete\n",
      "\n",
      "33 4DNESUNOW1OZ\n",
      "None human\n",
      "4DNEX1HYSKMF part1 complete\n",
      "4DNEXMFJRLJN part1 complete\n",
      "\n",
      "34 4DNESJPG5AS9\n",
      "None human\n",
      "4DNEXZLZZO9L part1 complete\n",
      "4DNEXHV4R6RJ part1 complete\n",
      "\n",
      "35 4DNESCOUGP8R\n",
      "None human\n",
      "4DNEX61WVI1Y part1 complete\n",
      "4DNEXY72J4DD part1 complete\n",
      "\n",
      "36 4DNES4CEYDPH\n",
      "None human\n",
      "4DNEXGRP461A part1 complete\n",
      "4DNEX8I12H9D part1 complete\n",
      "\n",
      "37 4DNES8128EIS\n",
      "None human\n",
      "4DNEXMHCH1TT part1 complete\n",
      "4DNEXJFQRJ38 part1 complete\n",
      "\n",
      "38 4DNESUJC9Y83\n",
      "None human\n",
      "4DNEXN62FHTX part1 complete\n",
      "4DNEXSA3QR2O part1 complete\n",
      "\n",
      "39 4DNES4B7RLAV\n",
      "None human\n",
      "4DNEX6GMB44Z part1 complete\n",
      "4DNEXWR7XY8S part1 complete\n",
      "\n",
      "40 4DNESDQ9PZOX\n",
      "None human\n",
      "4DNEXFBHF2ZI part1 complete\n",
      "4DNEXGZEOJE6 part1 complete\n",
      "\n",
      "41 4DNESO83H9ZI\n",
      "None human\n",
      "4DNEXF2KKZF1 part1 complete\n",
      "4DNEXI55T28T part1 complete\n"
     ]
    }
   ],
   "source": [
    "import time\n",
    "\n",
    "# for a given experiment set and some parameters like instrument\n",
    "# print set of files and their partA hic workflow status\n",
    "# if there are one that are running report the number of running cases\n",
    "# if there are file pairs that don't have a corresponding part A, report them separately\n",
    "\n",
    "out_n = \"This is an output file of the RepliSeq processing pipeline\"\n",
    "int_n = \"This is an intermediate file in the Repliseq processing pipeline\"\n",
    "\n",
    "def step_settings(seq, my_organism):\n",
    "    genome = \"\"\n",
    "    mapper = {'human':'GRCh38','mouse':'GRCm38'}\n",
    "    genome = mapper.get(my_organism)\n",
    "\n",
    "    wf_dict =[{\n",
    "        'wf_name': 'repliseq-parta',\n",
    "        'wf_uuid': '146da22a-502d-4500-bf57-a7cf0b4b2364',\n",
    "        \"parameters\" :  { \"nthreads\": 4, \"memperthread\": \"2G\" },\n",
    "        'custom_pf_fields': {\n",
    "            'filtered_sorted_deduped_bam': {\n",
    "                'genome_assembly': genome,\n",
    "                'file_type': 'alignment',\n",
    "                'description': out_n},\n",
    "            'count_bg': {\n",
    "                'genome_assembly': genome,\n",
    "                'file_type': 'read counts',\n",
    "                'description': out_n}\n",
    "        }}]\n",
    "    return wf_dict[seq] \n",
    "    \n",
    "# url for hic exps\n",
    "exp_types = ['Repli-seq']\n",
    "#exp_types = ['DAM-ID+seq']\n",
    "\n",
    "set_url = '/search/?'+'&'.join(['experiments_in_set.experiment_type='+i for i in exp_types])+'&type=ExperimentSetReplicate&limit=all'\n",
    "run_sets = ff_utils.search_metadata(set_url ,ff_env = env)\n",
    "\n",
    "add_pc = False\n",
    "add_rel = False\n",
    "add_wfr = False\n",
    "\n",
    "counter = 0\n",
    "completed = 0\n",
    "completed_acc = []\n",
    "all_sets = len(run_sets)\n",
    "print(str(all_sets)+' total number of sets')\n",
    "\n",
    "run_sets = [i for i in run_sets if \"RepliSeq_Pipeline_0.2.5\"  not in i.get('completed_processes', [])]\n",
    "\n",
    "print(str(all_sets-len(run_sets))+ ' sets completed')\n",
    "\n",
    "\n",
    "for a_set in run_sets: \n",
    "    counter += 1\n",
    "    print\n",
    "    fastqpairs, organism, enzyme, bwa_ref, chrsize_ref, enz_ref, f_size = find_pairs(a_set, exclude_miseq, env, tibanna, lookfor = 'single')\n",
    "    if not bwa_ref or not chrsize_ref:\n",
    "        print counter, a_set['accession'], organism, enzyme, 'skipping set with not chrsize/bwa index'\n",
    "        continue  \n",
    "    print counter, a_set['accession']\n",
    "    print enzyme, organism\n",
    "    # part3 = 'done'\n",
    "    list_release = []\n",
    "    # set_pairs = []        \n",
    "    # cycle through the experiments\n",
    "    for exp in fastqpairs.keys():\n",
    "        if not fastqpairs.get(exp):\n",
    "            print(exp, 'does not have any fastq pairs')\n",
    "            continue\n",
    "        # Check Part 1 and See if all are okay\n",
    "        exp_bams = []\n",
    "        part1 = 'done'\n",
    "        # part2 = 'done'\n",
    "        for pair in fastqpairs[exp]:\n",
    "            #############\n",
    "            bam = get_wfr_out(pair, 'repliseq-parta 0.2.5', 'bam', env)\n",
    "            # if run is not successful\n",
    "            if bam.startswith('no') or not bam:\n",
    "                part1 = 'not ready'\n",
    "                if add_wfr:\n",
    "                    inp_f = {'fastq':pair, 'chromsizes':chrsize_ref, 'bwaIndex':bwa_ref}\n",
    "                    name_tag = pair.split('/')[2]\n",
    "                    run_missing_wfr(step_settings(0, organism), inp_f, name_tag, env, tibanna)\n",
    "            elif bam == 'running':\n",
    "                part1 = 'still running'\n",
    "                print('part1 still running')\n",
    "            # if successful\n",
    "            else:\n",
    "                exp_bams.append(bam)\n",
    "                list_release.append(bam)\n",
    "        # stop progress to part2 \n",
    "        if part1 is not 'done':\n",
    "            print exp, 'has missing Part1 runs'\n",
    "            part2 = 'not ready'\n",
    "            part3 = 'not ready'\n",
    "            continue\n",
    "        print exp, 'part1 complete'\n",
    "\n",
    "\n",
    "# "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "        #check if part 2 is run already, it not start the run\n",
    "#         exp_com_bam = []\n",
    "#         exp_pairs = []\n",
    "#         for bam in exp_bams:\n",
    "#             com_bam = get_wfr_out(bam, 'hi-c-processing-bam 0.2.5', 'bam', ff)\n",
    "#             pairs = get_wfr_out(bam, 'hi-c-processing-bam 0.2.5', 'pairs', ff)\n",
    "#             # try to run if missing\n",
    "#             if pairs.startswith('no') or not pairs:\n",
    "#                 part2 = 'not ready'\n",
    "#                 part3 = 'not ready'\n",
    "                \n",
    "#             elif pairs == 'running':\n",
    "#                 print(bam)\n",
    "#                 part2 = 'still running'\n",
    "#                 part3 = 'not ready'\n",
    "                \n",
    "#             else:\n",
    "#                 exp_com_bam.append(com_bam)\n",
    "#                 exp_pairs.append(pairs)\n",
    "                \n",
    "#         # if still running, skip to next experiment\n",
    "#         if part2 == 'still running':\n",
    "#             print('part2 still running')\n",
    "#             continue\n",
    "        \n",
    "#         # make sure all bams went through the same wfr and produces same file\n",
    "#         if part2 != 'done' or len(list(set(exp_com_bam))) != 1 or len(list(set(exp_pairs))) !=1:\n",
    "#             print exp, 'Part2 did not complete'\n",
    "#             part3 = 'not ready' \n",
    "        \n",
    "#             if add_wfr:\n",
    "#                 if not chrsize_ref:\n",
    "#                     print 'not yet usable', organism\n",
    "#                     continue\n",
    "#                 # make sure no duplicates\n",
    "#                 inp_f = {'input_bams':exp_bams, 'chromsize':chrsize_ref}           \n",
    "#                 run_missing_wfr(wf_dict[1], inp_f, exp, ff, env, tibanna)   \n",
    "#             continue\n",
    "            \n",
    "#         # add bam and pairs to exp proc file\n",
    "#         list_release.extend([exp_com_bam[0],exp_pairs[0]])\n",
    "#         if add_pc:\n",
    "#             add_processed_files(exp, [exp_com_bam[0],exp_pairs[0]], ff)\n",
    "        \n",
    "#         print exp, 'part2 complete'\n",
    "#         set_pairs.append(exp_pairs[0])\n",
    "    \n",
    "#     if part3 != 'done':\n",
    "#         print 'Part3 not ready'\n",
    "#         continue\n",
    "    \n",
    "#     if not set_pairs:\n",
    "#         print 'no pairs can be produced from this set'\n",
    "#         continue\n",
    "        \n",
    "#     merged_pairs = []\n",
    "#     for set_pair in set_pairs:\n",
    "#         merged_pair = get_wfr_out(set_pair, 'hi-c-processing-pairs 0.2.5', 'pairs', ff)\n",
    "#         hic = get_wfr_out(set_pair, 'hi-c-processing-pairs 0.2.5', 'hic', ff)\n",
    "#         mcool = get_wfr_out(set_pair, 'hi-c-processing-pairs 0.2.5', 'mcool', ff)\n",
    "#         normvec = get_wfr_out(set_pair, 'hi-c-processing-pairs 0.2.5', 'normvector_juicerformat', ff)\n",
    "#         if merged_pair.startswith('no') or not merged_pair:\n",
    "#             part3 = 'not ready'\n",
    "#             break\n",
    "#         elif merged_pair == 'running':\n",
    "#             part3 = 'still running'\n",
    "#             break\n",
    "#         else:\n",
    "#             merged_pairs.append(merged_pair)\n",
    "    \n",
    "    \n",
    "#     # if part3 is still running report it, and skip the rest of the script\n",
    "#     if part3 == 'still running':\n",
    "#         print 'part3', part3\n",
    "#         continue        \n",
    "                \n",
    "#     if part3 != 'done' or len(list(set(merged_pairs))) != 1:\n",
    "#         print a_set['accession'], 'is missing Part3'\n",
    "        \n",
    "#         # if it is not run, and add_wfr is true, go for it, then skip the rest of the script\n",
    "#         if add_wfr:\n",
    "#             if not chrsize_ref:\n",
    "#                 print 'not yet usable', organism\n",
    "#                 continue\n",
    "\n",
    "#             if not enz_ref:\n",
    "#                 print 'restriction enzyme not ready for', organism, enzyme\n",
    "#                 continue\n",
    "#             inp_f = {'input_pairs':set_pairs, 'chromsizes':chrsize_ref, 'restriction_file': enz_ref} \n",
    "#             run_missing_wfr(wf_dict[2], inp_f, a_set['accession'], ff, env, tibanna)\n",
    "#         continue\n",
    "#     #####\n",
    "#     #add competed flag to experiment\n",
    "#     if add_pc and add_rel:\n",
    "#         ff_utils.patch_metadata({\"completed_processes\":[\"RepliSeq_Pipeline_0.2.5\"]}, obj_id=a_set['accession'] ,connection=ff)\n",
    "    \n",
    "#     # add processed files to set\n",
    "#     list_release.extend([merged_pair, hic, mcool, normvec])\n",
    "#     if add_pc:\n",
    "#         add_processed_files(a_set['accession'], [merged_pair, hic, mcool, normvec], ff)\n",
    "    \n",
    "#     #release files and wfrs\n",
    "#     if add_rel:\n",
    "#         release_files(a_set['accession'], list(set(list_release)), ff)\n",
    "    \n",
    "#     completed += 1\n",
    "#     completed_acc.append(a_set['accession'])\n",
    "#     print a_set['accession'], 'part3 complete'\n",
    "\n",
    "    \n",
    "# print completed\n",
    "# print completed_acc"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "anaconda-cloud": {},
  "kernelspec": {
   "display_name": "Python 2",
   "language": "python",
   "name": "python2"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 2
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython2",
   "version": "2.7.10"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
